Managed Identity Documentation (#6763)
AzureAD/microsoft-authentication-library-for-js: lib/msal-node/docs/managed-identity.md

# Managed Identity with MSALJS

### Note

> > This feature is not yet officially supported in MSALJS. An [alpha package](https://www.npmjs.com/package/@azure/msal-node/v/2.3.0-alpha.0) has been upload to npm.

Reviewer: ```suggestion
> This feature is in preview and feedback is welcome. See the [preview package](https://www.npmjs.com/package/@azure/msal-node/v/2.3.0-alpha.0)
```

> A common challenge for developers is the management of secrets, credentials, certificates, and keys used to secure communication between services. [Managed identities](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/overview) in Azure eliminate the need for developers to handle these credentials manually. MSALJS supports acquiring tokens through the managed identity service when used with applications running inside Azure infrastructure, such as:

Reviewer: ```suggestion
A common challenge for developers is the management of secrets, credentials, certificates, and keys used to secure communication between services. [Managed identities](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/overview) in Azure eliminate the need for developers to handle these credentials manually. MSALJS for Node supports acquiring tokens through the managed identity service when used with applications running inside Azure infrastructure, such as:
```

> 

Reviewer: Pls add a note here that clarifies that only the Node version of MSALJS supports confidential client and managed identity. Browser-based MSAL JS does not offer confidential flows because there is no confidentiality between the browser and the token issuer and does not offer managed identity because the browser is not hosted on Azure.

-   [Azure App Service](https://azure.microsoft.com/products/app-service/) (API version `2019-08-01` and above)
-   [Azure VMs](https://azure.microsoft.com/free/virtual-machines/)
-   [Azure Arc](https://learn.microsoft.com/en-us/azure/azure-arc/overview)

> -   [Azure Cloud Shell](https://learn.microsoft.com/en-us/azure/cloud-shell/overview)

Reviewer: Pls add a note taht Cloud Shell and Service Fabric are not yet supported.

### System-assigned managed identities

For system-assigned managed identities, the developer does not need to pass any additional information when creating an instance of ManagedIdentityApplication, as it will automatically infer the relevant metadata about the assigned identity.

[acquireToken](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/msi_feature_branch/lib/msal-node/src/client/ManagedIdentityApplication.ts#L122) is called with the resource to acquire a token for, such as `https://management.azure.com`.

```typescript
// optional
const config: ManagedIdentityConfiguration = {
    system: {
        loggerOptions: {
            logLevel: LogLevel.Verbose,
            loggerCallback: (_level, message, _containsPii) => {
                console.log(message);
            },
            piiLoggingEnabled: true,
        },
    },
};

const systemAssignedManagedIdentityApplication = new ManagedIdentityApplication(
    config
);

> const managedIdentityRequest: ManagedIdentityRequestParams = {

Reviewer: var name should be more like it's type

    resource: "https://management.azure.com",
};

try {
    const response =
        await systemAssignedManagedIdentityApplication.acquireToken(
            managedIdentityRequest
        );
    console.log(response);

> } catch (error) {

Reviewer: Simplify this code snippet by removing the try / catch.

Like in the case for system-assigned managed identities, ` acquireToken` is called with the resource to acquire a token for, such as `https://management.azure.com`.

```typescript
const managedIdentityIdParams: ManagedIdentityIdParams = {
    userAssignedClientId: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
};

const config: ManagedIdentityConfiguration = {
    managedIdentityIdParams,
    // optional
    system: {
        loggerOptions: {
            logLevel: LogLevel.Verbose,
            loggerCallback: (_level, message, _containsPii) => {
                console.log(message);
            },
            piiLoggingEnabled: true,
        },
    },
};

const userAssignedClientIdManagedIdentityApplication =
    new ManagedIdentityApplication(config);

const managedIdentityRequest: ManagedIdentityRequestParams = {
    resource: "https://management.azure.com",
};

try {
    const response =
        await userAssignedClientIdManagedIdentityApplication.acquireToken(
            managedIdentityRequest
        );
    console.log(response);
} catch (error) {
    console.log(error);
    throw error;
}
```

## Caching

> By default, MSALJS supports in-memory caching. Managed identity in MSALJS does not support cache extensibility because of security concerns when using distributed cache. Since a token acquired for managed identity belongs to an Azure resource, using a distributed cache might expose it to the other Azure resources sharing the cache.

Reviewer: ```suggestion
MSALJS caches tokens from managed identity in memory. There is no eviction, but memory is not a concern because there can only exist a limited number of managed identities defined. Cache extensibility is not supported in this scenario because tokens should not be shared between machines.
```
